### PR TITLE
Check for invalid optional arguments

### PR DESCRIPTION
**Pull request by luise**

### CODE REVIEW

File: stitch/bindings.js [hunk 1]
```diff
@@ -9,6 +9,7 @@ const os = require('os');
 const fs = require('fs');
```
Comment (kklin) [inline stitch/bindings.js]: This deserves to be in the changelog IMO

File: stitch/bindings.js [hunk 2]
```diff
@@ -439,6 +442,36 @@ function boxRange(x) {
...
 
 /**
+  * Throws an error if the first object contains any keys that do not appear in
+  * the second object. This is useful for checking if the user passed any
+  * invalid options to functions that take optional arguments.
+  * @private
+  *
+  * @param {Object} opts - The Object to check for redundant keys.
+  * @param {Object} obj - The object to compare against.
+  * @return {void} Throws an error if redundant keys are found.
```
Comment (kklin) [inline stitch/bindings.js]: I'm a bit surprised there's no JSDoc tag for errors

File: stitch/bindings.js [hunk 3]
```diff
@@ -439,6 +442,36 @@ function boxRange(x) {
...
+  */
+function checkExtraKeys(opts, obj) {
+  // Sometimes, prototype constructors are called internally by Quilt. In these
+  // cases, an existing object is passed as the optional argument, and the
+  // optional argument thus contains not just the keys  passed by the user, but
```
Comment (kklin) [inline stitch/bindings.js]: Two spaces here

File: stitch/bindings.js [hunk 4]
```diff
@@ -439,6 +442,36 @@ function boxRange(x) {
...
+  // also the keys Quilt set on the object, as well as all the prototype
+  // methods. Since we only want to check the optional arguments passed by the
+  // user, we skip all calls internally from Quilt (indicated by having the
+  // refID set in the options).
+  if (objectHasKey.call(opts, '_refID')) { return; }
+
+  const extras = [];
+  Object.keys(opts).forEach((key) => {
```
Comment (kklin) [inline stitch/bindings.js]: This would be a bit cleaner if you used `const extras = Object.keys(opts).filter(...)`

File: stitch/bindings_test.js [hunk 5]
```diff
@@ -82,6 +82,13 @@ describe('Bindings', () => {
         sshKeys: ['key1', 'key2'],
       }]);
     });
+    it('errors when passed invalid optional arguments', () => {
+      expect(() => new Machine({ badArg: 'foo' })).to
+        .throw('Unrecognized keys passed to Machine constructor: ');
```
Comment (kklin) [inline stitch/bindings_test.js]: Shouldn't there be the list of extra keys in all these `throw` messages?

File: stitch/bindings.js [hunk 6]
```diff
@@ -439,6 +442,36 @@ function boxRange(x) {
 }
 
 /**
+  * Throws an error if the first object contains any keys that do not appear in
```
Comment (kklin) [inline stitch/bindings.js]: I think it'd be helpful to explain here the pattern of checking the argument object against the object we instantiate.

Comment (kklin) [reply stitch/bindings.js]: The fact that we use that pattern, and why it's a valid approach.

File: stitch/bindings.js [hunk 7]
```diff
@@ -439,6 +442,36 @@ function boxRange(x) {
...
+  */
+function checkExtraKeys(opts, obj) {
+  // Sometimes, prototype constructors are called internally by Quilt. In these
+  // cases, an existing object is passed as the optional argument, and the
+  // optional argument thus contains not just the keys  passed by the user, but
+  // also the keys Quilt set on the object, as well as all the prototype
+  // methods. Since we only want to check the optional arguments passed by the
+  // user, we skip all calls internally from Quilt (indicated by having the
+  // refID set in the options).
+  if (objectHasKey.call(opts, '_refID')) { return; }
```
Comment (kklin) [inline stitch/bindings.js]: Sigh, this is ugly.. we probably shouldn't be passing instantiated objects as the optional argument. I'm not sure what to do here.

File: stitch/bindings.js (hunk 6 above)
Comment (luise) [reply stitch/bindings.js]: Ah, cool, I'll do that!

File: stitch/bindings.js (hunk 2 above)
Comment (luise) [reply stitch/bindings.js]: There is a `@throws` tag! I'll fix.

File: stitch/bindings.js (hunk 7 above)
Comment (luise) [reply stitch/bindings.js]: Yeah, I wasn't sure either how to restructure the `clone()` calls. Do you think I should fix this in this PR? If so, I'll think more about `clone()`.

File: stitch/bindings_test.js (hunk 5 above)
Comment (luise) [reply stitch/bindings_test.js]: In the actual error message, yes, the extra keys will be there. But the test passes as long as the string is a substring of the error message -- and this way we don't have to deal with the indeterministic order of the keys.
Would you prefer having the keys in the test?

Comment (kklin) [reply stitch/bindings_test.js]: I'm fine leaving them out. What if you tested the full error for the errors that are only one argument?

File: stitch/bindings.js (hunk 7 above)
Comment (kklin) [reply stitch/bindings.js]: No, definitely not a fix for this PR

File: stitch/bindings_test.js (hunk 5 above)
Comment (luise) [reply stitch/bindings_test.js]: Yup, that's a great idea.

**Comment by luise** [comment]
@kklin , updated

**Pull request merged**